File: web/src/app/modules/common/data.service.ts
import {Injectable} from '@angular/core';
import {BehaviorSubject} from 'rxjs';
import {GameProvider} from '../game-data/game-provider';
import {Game} from '../game-data/model/game';
import {WebsocketService} from './websocket.service';

@Injectable({
  providedIn: 'root'
})
export class DataService {

  private data: BehaviorSubject<Game> = new BehaviorSubject(GameProvider.createGame());
  currentData = this.data.asObservable();

  constructor(private ws: WebsocketService) {
    ws.listen((game: Game) => {
      this.data.next(game);
    });
  }
}

File: core/src/main/java/eu/kotkas/villak/core/service/DataService.java
package eu.kotkas.villak.core.service;

import eu.kotkas.villak.core.model.Game;
import eu.kotkas.villak.core.model.Message;

/**
 * @author Kristen Kotkas
 */
public interface DataService {

  Game getNextState(Message message);

  Game reduce(Game game, Message message);

}
